### ISSUE TITLE
Tree not processing if the the tab is not active in WASM

### ISSUE THREAD

**Issue by Nolkaloid** [godotengine/godot#37031]
**Godot version:**
3.2.1
**OS/device including version:**
Firefox 73
**Issue description:**
When running a project in the browser, the tree pauses if you leave the tab. Timers are paused and no functions are fired. This is particularly an issue in online turn games when someone leaves the tab to check another website while waiting for their turn.

**Comment by Calinou** [comment]
cc @Faless

**Comment by Faless** [comment]
I don't know what we can do anything about that.
`requestAnimationFrame` (which is needed for proper rendering sync) is not called by browsers when the tab is not selected.
Using `sertTimeout` will solve that, but will most likely yield worst performances in the rendering part:

https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_set_main_loop_timing

https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Inactive_tabs

**Comment by danilw** [comment]
how is this bug?
Godot use `requestAnimationFrame` method to animate at browser max FPS, that do not work when tab hidden.
Using any other looping in Webbrowser to WebGL is broken by design (no Vsync, FPS unpredictable)

marking bugreports like this as *Bug* make me confused very much

**Comment by Faless** [comment]
Yeah, I agree, this bug is still being properly triaged. But it's alright,
we opened a discussion to see if someone has ideas on a potential solution
(which I don't see) and it will be closed if, as I suspect there is no way
around it.

**Comment by Nolkaloid** [comment]
Would it be possible to switch methods on the run in order to keep some stuff running when the tab is unfocused? For example timers. Otherwise this make multiplayer turn games broken.

**Comment by danilw** [comment]
>Would it be possible to switch methods on the run in order to keep some stuff running when the tab is unfocused? For example timers. Otherwise this make multiplayer turn games broken.

current Web-pages model is *execute it when it visible* only events that send from *browser* when tab not focused will work, like video/audio events for player-control-elements

in current state of WASM executing something additional is impossible(or requiare way too much work to do that)
WASM do not support multithreading, so there no way to use two threads where one for render other for logic and sync them...

only way you can do it(if want) is - create some dirt-javascript async hack that execute on its own(or call some other WASM module) when tab not focused and when tab become visible that javascript-function send data to your main WASM godot game... as you see this logic does not look good

**Comment by Nolkaloid** [comment]
I don't quite understand what you mean... Another thing that could be done is some signal that could be fired when you refocus the tab, so it can request an update from a server for example.

**Comment by danilw** [comment]
WASM not support that....
WASM has only one thread that executed by calling requestAnimationFrame every frame
that all

**Comment by Nolkaloid** [comment]
I know that WASM has only one thread, my idea was to switch from `RequestAnimationFrame` to `settTimeout` **when the tab is unfocused** and when the tab is focused (again) **switch back to** `RequestAnimationFrame`.

**Comment by danilw** [comment]
Error: WebGL lost

**Comment by Faless** [comment]
> I know that WASM has only one thread, my idea was to switch from `RequestAnimationFrame` to `settTimeout` **when the tab is unfocused** and when the tab is focused (again) **switch back to** `RequestAnimationFrame`.

Yeah, that could work but will have to be tested, in particular, we well need to make sure that during those callbacks happening in unfocused window/tab we don't call any OpenGL function.
Again, will have to be tested, but *in theory*, it could work.

**Comment by danilw** [comment]
>Yeah, that could work but will have to be tested

this will work only on PC Chrome webbrowser, when Windows energy-save profile set to "max erformance", even firefox and other browsers flag `settTimeout` as deprecated and fully stop javascript code from executing on hiden tabs. Mobile/MAC/other OS broesers stop tabs without any exception.

I do not like see this change in WebGL render, if it will be supported in 4.0
I know lots of problems with WebGL lost context when something executed in render loop outside of `RequestAnimationFrame`

**Comment by Faless** [comment]
> firefox and other browsers flag `settTimeout` as deprecated

Do you have a reference for that? This is the first time I hear `settTimeout` has been deprecated, and can't find any mention of it around....

> I know lots of problems with WebGL lost context when something executed in render loop outside of `RequestAnimationFrame`

As I explained above, the idea is **not to run the render loop at all** when not inside the animation frame.

**Comment by danilw** [comment]
https://emscripten.org/docs/api_reference/emscripten.h.html

>Browsers heavily optimize towards using requestAnimationFrame for animation instead of the other provided modes. Because of that, for best experience across browsers, calling this function with mode=EM_TIMING_RAF and value=1 will yield best results. Using the JavaScript setTimeout function is known to cause stutter and generally worse experience than using the requestAnimationFrame function.

>There is a functional difference between setTimeout and requestAnimationFrame: If the user minimizes the browser window or hides your application tab, browsers will typically stop calling requestAnimationFrame callbacks, but setTimeout-based main loop will continue to be run, although with heavily throttled intervals. See setTimeout on MDN <https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers.setTimeout#Inactive_tabs> for more information.

**Comment by Faless** [comment]
@danilw yeah, no mention of deprecation, no mention of not being supported, and the MDN docs actually says that throttling will not take place if and AudioContext is active.

**Comment by ignasKavaliauskas** [comment]
Stumbled upon this issue as well, happy to see it in the 4.0 milestone! Biggest issue with that in my case is that the networking is not being processed when out of focus. Once back in focus, it processes all the packets (+all the ones received during out of focus) but it may lead to unwanted issues related to the game state.

**Comment by Nolkaloid** [comment]
Exactly my issue, I abandoned development of my project because of this.

**Comment by Faless** [comment]
Please don't add comments if you don't have relevant information to add.

Use the "thumbs up" :+1: reaction in the top post to express your interest in the issue.

**Comment by ChronoDK** [comment]
I don't have new info, but my game was unplayable if the player minimized the window for a few seconds and then resumed. Many methods were never called, because they had been delayed using yield or tween. So I wrote this small workaround. Adding it here, in case it can help anyone.

In my global autoload called Game:

```
var delayedCalls = []
func delayedCall(delay, object, method, vars):
    delayedCalls.push_front({"delay":OS.get_system_time_msecs() + (delay * 1000), "object":object, "method":method, "vars":vars})
    pass
```

In the process loop of the active scene:

```
func _process(delta):
    for i in range(Game.delayedCalls.size()-1, -1, -1):
        var dc = Game.delayedCalls[i]
        if dc.delay <= OS.get_system_time_msecs():
            dc.object.callv(dc.method, dc.vars)
            Game.delayedCalls.remove(i)
```

And then from anywhere in the code you can call methods like this:

```
Game.delayedCall(delay, label, "set", ["visible", true])
Game.delayedCall(delay, label, "set", ["bbcode_text", "[center][wave amp=20 freq=10]HELLO[/wave][/center]"])
Game.delayedCall(delay, dialog, "show", [])
```
